File: contents/flood_fill/code/rust/flood_fill.rs
use ndarray::prelude::*; // 0.13.1
use std::collections::VecDeque;

fn color(canvas: &mut Array<usize, Ix2>, loc: &[usize; 2], old_val: usize, new_val: usize) {
    // do not color a nonexistent pixel
    if let Some(pixel) = canvas.get(*loc) {
        // only change color if the element value is the old value
        if pixel == &old_val {
            canvas[*loc] = new_val;
        }
    }
}

fn find_neighbors(canvas: &Array<usize, Ix2>, loc: &[usize; 2], val: usize) -> Vec<[usize; 2]> {
    // find neighbors
    let mut possible_neighbors: Vec<[usize; 2]> = vec![[loc[0] + 1, loc[1]], [loc[0], loc[1] + 1]];
    // guard against underflow errors
    if loc[0] as isize > 0 {
        possible_neighbors.push([loc[0] - 1, loc[1]]);
    }
    if loc[1] as isize > 0 {
        possible_neighbors.push([loc[0], loc[1] - 1]);
    }

    // exclude neighbors that are out of bounds and are not the old fill value
    possible_neighbors
        .into_iter()
        .filter(|i| canvas.get(*i) == Some(&val))
        .collect()
}

fn stack_fill(canvas: &mut Array<usize, Ix2>, loc: &[usize; 2], old_val: usize, new_val: usize) {
    // don't do anything if old_val and new_val are identical
    if new_val == old_val {
        return;
    }

    // set up the stack
    let mut stack: Vec<[usize; 2]> = Vec::new();
    stack.push(*loc);

    while let Some(current_loc) = stack.pop() {
        if canvas[current_loc] == old_val {
            color(canvas, &current_loc, old_val, new_val);
            let neighbors = find_neighbors(canvas, &current_loc, old_val);
            stack.extend(neighbors);
        }
    }
}

fn queue_fill(canvas: &mut Array<usize, Ix2>, loc: &[usize; 2], old_val: usize, new_val: usize) {
    // don't do anything if old_val and new_val are identical
    if new_val == old_val {
        return;
    }

    // set up the queue
    let mut queue: VecDeque<[usize; 2]> = VecDeque::new();
    queue.push_back(*loc);

    // color the initial location
    color(canvas, &loc, old_val, new_val);

    while let Some(current_loc) = queue.pop_front() {
        let up_next = find_neighbors(canvas, &current_loc, old_val);
        // color as we enqueue neighbors
        for n in up_next {
            color(canvas, &n, old_val, new_val);
            queue.push_back(n);
        }
    }
}

fn recursive_fill(
    canvas: &mut Array<usize, Ix2>,
    loc: &[usize; 2],
    old_val: usize,
    new_val: usize,
) {
    // don't do anything if old_val and new_val are identical
    if new_val == old_val {
        return;
    }

    color(canvas, &loc, old_val, new_val);

    let up_next = find_neighbors(canvas, &loc, old_val);
    up_next
        .iter()
        .for_each(|i| recursive_fill(canvas, i, old_val, new_val));
}

fn main() {
    let grid = array![
        [0, 0, 1, 0, 0],
        [0, 0, 1, 0, 0],
        [0, 0, 1, 0, 0],
        [0, 0, 1, 0, 0],
        [0, 0, 1, 0, 0]
    ];

    let solution_grid = array![
        [1, 1, 1, 0, 0],
        [1, 1, 1, 0, 0],
        [1, 1, 1, 0, 0],
        [1, 1, 1, 0, 0],
        [1, 1, 1, 0, 0]
    ];

    let start_loc = [0, 0];

    let mut recursive_grid = grid.clone();
    recursive_fill(&mut recursive_grid, &start_loc, 0, 1);

    let mut queue_grid = grid.clone();
    queue_fill(&mut queue_grid, &start_loc, 0, 1);

    let mut stack_grid = grid.clone();
    stack_fill(&mut stack_grid, &start_loc, 0, 1);

    println!("Starting grid:");
    println!("{}", grid);
    println!();
    println!("Solution grid:");
    println!("{}", solution_grid);
    println!();
    println!(
        "Recursive grid {} correct.",
        if recursive_grid == solution_grid {
            "is"
        } else {
            "is NOT"
        }
    );
    println!(
        "Queue grid {} correct.",
        if queue_grid == solution_grid {
            "is"
        } else {
            "is NOT"
        }
    );
    println!(
        "Stack grid {} correct.",
        if stack_grid == solution_grid {
            "is"
        } else {
            "is NOT"
        }
    );
}

File: contents/flood_fill/code/julia/flood_fill.jl
using DataStructures
using Test

# Function to check to make sure we are on the canvas
function inbounds(canvas_size, loc)

    # Make sure we are not beneath or to the left of the canvas
    if minimum(Tuple(loc)) < 1
        return false

    # Make sure we are not to the right of the canvas
    elseif loc[2] > canvas_size[2]
        return false

    # Make sure we are not above the canvas
    elseif loc[1] > canvas_size[1]
        return false
    else
        return true
    end
end

function find_neighbors(canvas, loc::CartesianIndex, old_val, new_val)
    
    # Finding north, south, east, west neighbors
    possible_neighbors = [loc + CartesianIndex(0, 1),
                          loc + CartesianIndex(1, 0),
                          loc + CartesianIndex(0, -1),
                          loc + CartesianIndex(-1, 0)]

    # Exclusing neighbors that should not be colored
    neighbors =  []
    for possible_neighbor in possible_neighbors
        if inbounds(size(canvas), possible_neighbor) &&
           canvas[possible_neighbor] == old_val
            push!(neighbors, possible_neighbor)
        end
    end

    return neighbors
end

function stack_fill!(canvas, loc::CartesianIndex, old_val, new_val)
    if new_val == old_val
        return
    end

    s = Stack{CartesianIndex}()
    push!(s, loc)

    while length(s) > 0
        current_loc = pop!(s)
        if canvas[current_loc] == old_val
            canvas[current_loc] = new_val
            possible_neighbors = find_neighbors(canvas, current_loc,
                                                old_val, new_val)
            for neighbor in possible_neighbors
                push!(s,neighbor)
            end
        end
        
    end
end


function queue_fill!(canvas, loc::CartesianIndex, old_val, new_val)
    if new_val == old_val
        return
    end

    q = Queue{CartesianIndex}()
    enqueue!(q, loc)

    # Coloring the initial location
    canvas[loc] = new_val

    while length(q) > 0
        current_loc = dequeue!(q)

        possible_neighbors = find_neighbors(canvas, current_loc,
                                            old_val, new_val)

        # Coloring as we are enqueuing neighbors
        for neighbor in possible_neighbors
            canvas[neighbor] = new_val
            enqueue!(q,neighbor)
        end
        
    end
end

function recursive_fill!(canvas, loc::CartesianIndex, old_val, new_val)

    if (old_val == new_val)
        return
    end

    canvas[loc] = new_val

    possible_neighbors = find_neighbors(canvas, loc, old_val, new_val)
    for possible_neighbor in possible_neighbors
        recursive_fill!(canvas, possible_neighbor, old_val, new_val)
    end
end

function main()

    # Creation of a 5x5 grid with a single row of 1.0 elements 
    grid = zeros(5,5)
    grid[3,:] .= 1

    # Create solution grid
    answer_grid = zeros(5,5)
    answer_grid[1:3, :] .= 1

    # Start filling at 1,1
    start_loc = CartesianIndex(1,1)

    @testset "Fill Methods" begin
        # Use recursive method and reinitialize grid
        recursive_fill!(grid, start_loc, 0.0, 1.0)
        @test grid == answer_grid

        grid[1:2,:] .= 0

        # Use queue method and reinitialize grid
        queue_fill!(grid, start_loc, 0.0, 1.0)
        @test grid == answer_grid

        grid[1:2,:] .= 0

        # Use stack method and reinitialize grid
        stack_fill!(grid, start_loc, 0.0, 1.0)
        @test grid == answer_grid
    end

end

main()
